### ISSUE TITLE
Configurable K8S probes

### ISSUE THREAD

**Issue by Alexx-G** [coreos/prometheus-operator#3257]
**What happened?**

In case the volume of data is huge (e.g. 150+ pods, each with ~4-5MB of metrics), Prometheus doesn't manage to start within liveness probe limits. It just stucks in a restart loop, trying to load WAL and getting killed by K8S.

**Did you expect to see something different?**

I'd expect liveness probe to be either configurable or at least comparable to readiness probe, which is huge.
Allowing to set startupProbe would also help.

**How to reproduce it (as minimally and precisely as possible)**:

In our case it just happens once amount of data gets huge and prometheus pod gets bounced and tries to load data.
Everything is provisioned using prometheus-operator helm chart with thanos sidecar.

**Environment**

* Prometheus Operator version:

    `v0.36.0`
    

* Kubernetes version information:

    `1.14`

* Kubernetes cluster kind: AWS EKS

**Comment by brancz** [comment]
Could you create a PR to set liveness equally large as readiness?

**Comment by brancz** [comment]
FWIW I even wouldn't mind removing the liveness probe all together, I feel it's causing more harm than use.

**Comment by Alexx-G** [comment]
I'll give it a try, though I'm not familiar with the codebase - just setting lower expectations.
I think it's better to have liveness probe than not to have it, but getting a right limit is tough. While 5min would seem more than enough, it appears that for some use cases it's not enough.
Setting a large liveness probe also has its own downsides - in case there's a glitch or huge slowness in prometheus (e.g. due to GC slowness), means that Promethes would become not ready / restarted after 25 min.
It might make sense to actually allow configuring [startupProbe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) (setting some reasonable defaults) - it would ensure that liveness probe isn't abused with huge values.

**Comment by brancz** [comment]
Yeah I'm happy with startupProbe as well. For what it's worth Prometheus is working on drastic improvements on WAL replay time, which even for huge setups should get it down to only a few minutes.

**Comment by Alexx-G** [comment]
Would you mind if I submit a PR in a few days with startup probe similar to liveness probe? In this case it’s possible to leave liveness/readiness probes untouched

**Comment by brancz** [comment]
that would be awesome!

**Comment by inoam** [comment]
+1 for configurable probes.
I would also like a configurable probe, and would like it to liveness. 
I have different case:
Suggest storage system has some glitch, it might take more than 30 seconds (6 retries * 5 seconds) for block device to reconnect.
In the current code prometheus would be restarted, but as said here, starup could take quite long (minutes). In this case small glitch of 30 seconds in storage would end up in 3 minutes of now k8s.

@Alexx-G do you plan to add configurable liveness probe, or would you add only startup ?

**Comment by Alexx-G** [comment]
I’ve looked into it and it might be a bit more complicated than I expected. startupProbes have been introduced in K8S 1.16 and IIRC they’re still in alpha.
Thus I’d really do 2 things - making liveness probe less strict (doubling retries) and in case K8S version is >= 1.16, add startup probes.

TBH, it might make sense to make them configurable, but afaik it would mean updating CRDs, is this right?

**Comment by brancz** [comment]
I would prefer not to add fields for a short team workaround. Both the Prometheus project as well as Kubernetes are working on features or have already released features to resolve this (for Prometheus this is WAL snapshots, for Kubernetes startup probes). For a short term workaround, I think people should override the liveness probe by specifying an alternative probe in via the `containers` array (which in an unsupported feature allows you to arbitrarily modify the prometheus container specification).
